core/src/services/upyun/core.rs

```diff
@@ -0,0 +1,557 @@
+    match headers.get(X_UPYUN_MULTI_UUID) {
+        None => Err(Error::new(ErrorKind::Unexpected, "missing uuid")),
+        Some(v) => Ok(v.to_str().map_err(|e| {
+            Error::new(
+                ErrorKind::Unexpected,
+                "header value has to be valid utf-8 string",
+            )
+            .with_operation("parse_initiate_part")
+            .set_source(e)
+        })?),
+    }
+}
+
+pub(super) fn parse_info(headers: &HeaderMap) -> Result<Metadata> {
+    let mode = if parse_file_type(headers)? == "file" {
+        EntryMode::FILE
+    } else {
+        EntryMode::DIR
+    };
+
+    let mut m = Metadata::new(mode);
+
+    if let Some(v) = parse_file_size(headers)? {
+        m.set_content_length(v);
+    }
+
+    if let Some(v) = parse_content_type(headers)? {
+        m.set_content_type(v);
+    }
+
+    if let Some(v) = parse_content_md5(headers)? {
+        m.set_content_md5(v);
+    }
+
+    if let Some(v) = parse_cache_control(headers)? {
+        m.set_cache_control(v);
+    }
+
+    if let Some(v) = parse_content_disposition(headers)? {
+        m.set_content_disposition(v);
+    }
+
+    Ok(m)
+}
+
+fn parse_file_type(headers: &HeaderMap) -> Result<&str> {
+    match headers.get(X_UPYUN_FILE_TYPE) {
+        None => Err(Error::new(ErrorKind::Unexpected, "missing file type")),
+        Some(v) => Ok(v.to_str().map_err(|e| {
+            Error::new(
+                ErrorKind::Unexpected,
+                "header value has to be valid utf-8 string",
+            )
+            .with_operation("parse_file_type")
+            .set_source(e)
+        })?),
+    }
+}
+
+fn parse_file_size(headers: &HeaderMap) -> Result<Option<u64>> {
+    match headers.get(X_UPYUN_FILE_SIZE) {
+        None => Ok(None),
+        Some(v) => Ok(Some(
+            v.to_str()
+                .map_err(|e| {
+                    Error::new(
+                        ErrorKind::Unexpected,
+                        "header value is not valid utf-8 string",
+                    )
+                    .with_operation("http_util::parse_content_length")
+                    .set_source(e)
+                })?
+                .parse::<u64>()
+                .map_err(|e| {
+                    Error::new(ErrorKind::Unexpected, "header value is not valid integer")
+                        .with_operation("http_util::parse_content_length")
+                        .set_source(e)
+                })?,
+        )),
+    }
+}
+
+fn parse_cache_control(headers: &HeaderMap) -> Result<Option<&str>> {
+    match headers.get(X_UPYUN_CACHE_CONTROL) {
+        None => Ok(None),
+        Some(v) => Ok(Some(v.to_str().map_err(|e| {
+            Error::new(
+                ErrorKind::Unexpected,
+                "header value has to be valid utf-8 string",
+            )
+            .with_operation("parse_cache_control")
+            .set_source(e)
+        })?)),
+    }
+}
+
+fn parse_content_disposition(headers: &HeaderMap) -> Result<Option<&str>> {
```

Comment:
How about adding a `opendal::raw::parse_header_to_str` so we don't need to repeat eror handling logic here? We can submit a new PR for this refactor.

Reply:
Okay, I'll add it in another PR

Reply:
Tracked at https://github.com/apache/incubator-opendal/issues/3793